kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/KafkaProxy.java

```diff
@@ -80,14 +83,26 @@ public List<Future<?>> shutdownGracefully() {
     private @Nullable EventGroupConfig serverEventGroup;
     private @Nullable Channel metricsChannel;
 
-    public KafkaProxy(PluginFactoryRegistry pfr, Configuration config) {
+    public KafkaProxy(PluginFactoryRegistry pfr, Configuration config, Set<Feature> features) {
         this.pfr = Objects.requireNonNull(pfr);
-        this.config = Objects.requireNonNull(config);
+        this.config = validate(config, features);
         this.virtualClusters = config.virtualClusterModel();
         this.adminHttpConfig = config.adminHttpConfig();
         this.micrometerConfig = config.getMicrometer();
     }
 
+    @VisibleForTesting
+    static Configuration validate(Configuration config, Set<Feature> features) {
+        Optional<Map<String, Object>> development = config.development();
+        if (development.isPresent() && !development.get().isEmpty()) {
+            if (!features.contains(Feature.LOAD_TEST_CONFIGURATION)) {
+                LOGGER.error("test-only configuration for proxy present, but loading test-only configuration not enabled: {}", development.get());
+                throw new IllegalConfigurationException("test-only configuration for proxy present, but loading test-only configuration not enabled");
+            }
```

Comment:
This feels a bit odd given you have `Feature.maybeLogWarning`. Why don't we add a flag to the feature like `TestOnly` and make this something like

```
var errors = features.filter(feature.supports(development.get()).findAll().toList();
log.error(errors);
throw new ICE(errors);
```

Reply:
alright I've refactored to something like this, the existing `Set<Feature>` didn't allow for this kind of implementation.

I've changed it so there is now a `Features` class that knows the `enabled/disabled` state for every Feature and the Feature is responsible for applying the validation logic. I think it's tidier than using set contains.